### ISSUE TITLE
Unit test presenters

### ISSUE THREAD

**Issue by IgorGanapolsky** [sockeqwe/mosby#68]
Hi,
I was wondering if you plan to demo unit testing (on the jvm) your presenters.  This would be good to test the RXJava Observables.  So far I just see instrumentation tests in Mosby samples.

Thanks,
Igor

**Comment by sockeqwe** [comment]
Maybe, but not in the near feature as I'm busier then I wanted to. Maybe I will write a blog post about that topic in fall. At the moment I have other things in mind with higher priority.

**Comment by IgorGanapolsky** [comment]
Cool, thx.  I think one the biggest benefits of MVP is being able to write unit tests.  I'm people would like to learn...

**Comment by jjvargas** [comment]
Also RoboElectric has released 3.0 :) , It would be a good moment to learn how to use it for the tests.

**Comment by sockeqwe** [comment]
That is great! Haven't noticed 3.0 release yet! thanks for the hint.

I think @IgorGanapolsky is misunderstanding something or not sure what he actually wants to test:

In MVP you have 3 layer: Model (Business Logic) , View (UI) and Presenter. Each layer can be testet independently. A concrete example: You have a local SQL database storing some kind of  items, having a Presenter to load those items and interact with a View (MvpLceView) which basically displays the list of Items in a RecyclerView:

How would I test that is:

1. Testing Model / BusinessLogic:  Write Unit test just for testing your Model / Business Logic like inserting items in the database works correctly, queuing items from database and so on.

2. Test Presenter: So the presenter is connected to the database. The database is already testet. So what to test in presenter? Here you should test if Presenter calls Views methods correctly and as expected like `view.showLoading()`, `view.showError()` etc. Let's assume `ItemsView extends MvpLceView`. `ItemsView` is a interface:
```java
@Test public void testLoadingItemsSuccessful(){
  
     // A mock view for testing
     MyView view = new MyView {

          public boolean showLoadingCalled = false;
          public boolean showErrorCalled = false;
          public boolean showContentCalled = false;
          public boolean setDataCalled = false;

          public void showLoading(boolean pullToRefresh) {
               showLoadingCalled = true;
          }

          public void setData(List<Item> items){
             showItemsCalled = true;
          }

         public void showError(Throwable t, boolean pullToRefresh){
              showErrorCalled = true;
         }

        public void showContent(){
           showContentCalled();
        }
      }

    // Presenter to test
    ItemsPresenter presenter = new ItemsPresenter();
    presenter.attachView(view);  // Set the mock view

    presenter.loadItems();

   // After having loaded items, we test if view methods gets invoked correctly
   assertTrue(view.showLoadingCalled);
   assertTrue(view.showContentCalled);
   assertTrue(view.setDataCalled);
   assertFalse(view.showErrorCalled);

}
```

This is how to test an presenter, you basically check if the interaction with the view is as expected. You should also test if an execpetion while loading items has been handled successfully, in other words has been `view.showError()` called instead of `view.showContent()`. You may also consider to test if `presenter.detachView()` really sets the reference to your view to null. That are the things you should test when talking about testing presenter. If you use one of Mosby's MvpLce Presenter you can assume that this is alredy tested for you and you don't have to test that again.


3. Testing View: Now you know that model is tested and presenter is tested. So in View you basically just test if your UI is displaying things as expected like this:

```java
@Test public void testShowingItems(){

     ItemsActivity view =  ... ;
     
     view.loadData();
     
     assertEquals(view.contentView.getVisibility(), VIEW.VISIBLE);
     assertEquals(view.errorView.getVisibility(), VIEW.GONE);
     assertEquals(view.loadingView.getVisibility(), VIEW.GONE);

}
```

It's pseudo code. just to give you an idea how to test things.


So I guess when @IgorGanapolsky  is asking me how to test Presenter (the title of this issue) he is actually asking of how to test model / business logic, don't you? And business logic can be testet as a pure java unit test. Presenter as well. only View test may require android instrumentation tests.

Hope that helps

**Comment by IgorGanapolsky** [comment]
@sockeqwe I was specifically wondering about testing Observables: whether they return valid objects or make the appropriate network calls that they are set up to.  But thanks for your lengthy explanation on testing Presenters with SQLite, I may use that approach in my personal app.

**Comment by sockeqwe** [comment]
Testing observables, is testing your business logic.

there is a `observable.toBlocking()` method you can use to transfrom your (async) observable into a blocking one and ask for the returning values 

OR

`TestSubscriber`
More info here: https://medium.com/ribot-labs/unit-testing-rxjava-6e9540d4a329

**Comment by IgorGanapolsky** [comment]
I am curious, in your example code above - what is MyView?  Is it an extension of <b>MVPView</b>.  If so, that code doesn't compile.  Those public fields - showLoadingCalled, showContentCalled - cannot be resolved.

**Comment by sockeqwe** [comment]
Its a custom view interface of a hypothetical app ... 

Somethink like [LoginView](https://github.com/sockeqwe/mosby/blob/master/sample-mail/src/main/java/com/hannesdorfmann/mosby/sample/mail/login/LoginView.java) of the mail sample

**Comment by TheRishka** [comment]
I've encountered another problem while trying to Unit Test presenters. I keep getting "Wanted but not invoked" error with Mockito. I've also tried to use MvpLcePresenterTest.java in testing package of Mosby. Got the same error. Does anybody have any clues how to use Mockito for testing Mosby Presenters?

**Comment by sockeqwe** [comment]
What exactly do you want to test? Which mocked method is not invoked, from
View?

TheRishka <notifications@github.com> schrieb am Do., 14. Jan. 2016, 11:52:

**Comment by TheRishka** [comment]
Right now I've added your MvpLcePresenterTest into my project. And I'm just trying to cal startLceTest() method of it.

```(java)
public class MyPresenterTest extends MvpLcePresenterTest<List<Object>,
        MyFragmentView, MyFragmentPresenter> {
    @Test
    public void testBase(){
        startTest(false);
    }

    @Override
    protected Class<MyFragmentView> getViewClass() {
        return MyFragmentView.class;
    }

    @Override
    protected MyFragmentPresenter createPresenter() {
        return new MyFragmentPresenter();
    }

    @Override
    protected void loadData(MyFragmentPresenter presenter, boolean pullToRefresh) {
        presenter.loadSomeRandomStuff();
    }
}
```

It fails with error on line 79 of MvpLcePresenterTest:
```
Wanted but not invoked:
MyFragmentView.showLoading(false);
-> at MvpLcePresenterTest.startTest(MvpLcePresenterTest.java:79)
Actually, there were zero interactions with this mock.
```

**Comment by sockeqwe** [comment]
`MvpLcePresenterTest` is used to test if you `MyFragmentPresenter` calls the corresponding `MvpLceView` methods like `view.showLoading()`, `view.showContent()` etc. as expected.

I guess your test is failing because your Presenter doesn't implement that "LCE" calls correctly (or as expected by `MvpLcePresenterTest`. 

your method `presenter.loadSomeRandomStuff();` should call:
1. view.showLoading(false);
2. load some data and then either call:
  - view.setData(data); and view.showContent();
  - or view.showError();

But your presenter doesn't seem to call that. Therefore the test is failing.
If you can't find the problem yourself, you could post the code of `MyFragmentPresenter` and I will try to help you ...

`

**Comment by TheRishka** [comment]
I'm not sure if the problem is in presenter... Here is my `presenter.loadSomeRandomStuff();`:
```(java)
    @VisibleForTesting
    public void loadSomeRandomStuff() {
        getView().stopRefreshing();
        List<Object> stuff = setMockArrayList(5);

        getView().showContent();
        if (stuff != null && stuff.size() > 0) {
            getView().setContentVisible(true);
            getView().setData(stuff);
        } else {
            getView().setContentVisible(false);
        }
    }
```
Still the same error. `MvpLcePresenterTest.java:79`

**Comment by sockeqwe** [comment]
Yeah, your problem is that you use the `MvpLcePresenterTest` which is meant to test a presenter that manages a `MvpLceView`.
BUT
from your presenter implementation I can say that this is not a `MvpLceView`! LCE stands for Loading-Content-Error. Let's have a look at `MvpLceView`:

```java
interface MvpLceView<M> extends MvpView {

  public void showLoading(boolean pullToRefresh);
  public void showContent();
  public void showError(Throwable e, boolean pullToRefresh);
  public void setData(M data);
}
```

The `MvpLcePresenterTest` is just a template to test a presenter that invokes all of this 4 methods on the view. So the presenter that can be tested with `MvpLcePresenterTest` is something like this:

```java
void loadSomeRandomStuff(){

     getView.showLoading(false);

     try {
         List<Object> stuff = setMockArrayList(5); // may throws an error
         getView().setData(stuff);
         getView().showContent();
     } catch( Exception e) {
         getView().showError(e, false);
    }

}
```

So you simply don't have a view that fullfills the idea of  `MvpLceView`. Therefore the `MvpLcePresenterTest` will fail. 
If you read the error message carefully you notice that error:
```
Wanted but not invoked:
MyFragmentView.showLoading(false);
```

As already said, the `MvpLcePresenterTest` assumes that `getView(),showLoading(false)` has been called. But your Presenter actually don't call that method.

TL;DR `MvpLcePresenterTest` is not suitable for the kind of View / Presenter you have. You have to write your own Test!

**Comment by TheRishka** [comment]
`MyFragmentView` does extend MvpLceView.
But methods are quite bad, like `stopRefreshing()` is useless coz `showLoading()`  is already defined. So I've redone testable presenter and `MyFragmentView`.
Now my method looks like this:
```(java)
void loadSomeRandomStuff(){
     getView.showLoading(false);
     try {
         List<Object> stuff = setMockArrayList(5); // may throws an error
         getView().setData(stuff);
         getView().showContent();
     } catch( Exception e) {
         getView().showError(e, false);
    }
}
```
Just like you told me. Now it throws another error! It says that some method wasn't wanted but invoked `showError`. So it's clear for me that it wasn't Mockito or Mosby problem! It was me who misunderstood Mockito messages! I can deal with it on my own, I hope:) Much thanks for your super-fast replies!

**Comment by sockeqwe** [comment]
This error comes from the fact that this test template also tests that
getView().showError() has been invoked ... Checkout the source code of
startLceTest() and you will see what's going on

TheRishka <notifications@github.com> schrieb am Do., 14. Jan. 2016, 14:42:

**Comment by TheRishka** [comment]
Ye, I've digged into it. The current problem now is that I'm testing presenter same presenter method, but on `setData()` it throws error `NotAMockException`.
The method is the same:
```(java)
void loadSomeRandomStuff(){
     getView.showLoading(false);
     try {
         List<Object> stuff = setMockArrayList(5); // may throws an error
         getView().setData(stuff);
         getView().showContent();
     } catch( Exception e) {
         getView().showError(e, false);
    }
}
```
It seems like I can't generate List<Object> that way. So I've tried to Mock List<Object> data and pass it as parameter:
```(java)
void loadSomeRandomStuff(List<Object> data){
     getView.showLoading(false);
     try {
         getView().setData(data);
         getView().showContent();
     } catch( Exception e) {
         getView().showError(e, false);
    }
}
```
That's the way I'm trying to pass objects:
```(java)

    @Mock
    private Object obj;
    
    @Before
    public void setUpMocks(){
        MockAnnotations.initMocks(this);
    }
    ....
    ....
    ....
    @Override
    protected void loadData(MyFragmentPresenter presenter, boolean pullToRefresh) {
        List<Object> data = Arrays.asList(obj, obj, obj);
        presenter.loadSomeRandomStuff(data);
    }
```

So...Again, I'm missing something. Would be great, if you could help me once again:(

**Comment by sockeqwe** [comment]
I think there is a misunderstanding on what the goal of testing a presenter is.

So may I ask you, what are you trying to test? What do you actually want to test?

**Comment by TheRishka** [comment]
I'm trying to test presenter method by method. I want to ensure that `loadSomeRandomStuff()` works as intended, data is displayed and loading set to false.

**Comment by sockeqwe** [comment]
alright, that sounds good.
But is `void loadSomeRandomStuff(List<Object> data)` really the method you will invoke in your app? To me it seems that you want to invoke this method just for testing (parameter data?)

So what you typically do is mocking the view and the business logic. With business logic I mean "the component that loads the data". 

Actually you are **not** testing the view (is something displayed, is View with id R.id.contentView == View.VISIBILE) and you are **not** testing if the business logic is working correctly (does the data is loaded successfully and containing the data x,y,z ). In MVP you test each layer independently from the other: 
- M = Business logic, i.e. Database query works correctly , loading data over http works and gets parsed correctly etc.
- V = View (i.e. UI tests with espresso): Does View with id R.id.contentView is actually displayed (view has visibility View.VISIBILE or View.GONE), does a click on a certain View like a button triggers an action etc.
- P = Presenter. here you just test if the Presenter is invoking the corresponding View methods correctly since the Presenter "manages/coordinates" the View so here you test something like this:
_"If presenter.loadData() is invoked and could not load data (i.e. simulate no active internet connection) then view.showError() should be called"
